### ISSUE TITLE
Rule proposal: require `key` prop for rendered array items

### ISSUE THREAD

**Issue by lencioni** [yannickcr/eslint-plugin-react#40]
I'm not sure how feasible this would be to implement, but it would be really great if we could catch this before runtime.

Good:

```jsx
<div>
  {myArray.map(item => <div key={item.id}>{item.body}</div>)}
</div>
```

Bad:

```jsx
<div>
  {myArray.map(item => <div>{item.body}</div>)}
</div>
```

**Comment by yannickcr** [comment]
Would be very useful but I am not sure how to do this. Need to investigate.

**Comment by benmosher** [comment]
I'm willing to take a stab at this.

I feel like at a minimum, array literals can be reported:

```js
<dl>
  items.map(item => [<dt>{item.name}</dt>, <dd>{item.desc}</dd>])
</dl>
```

More generally, I'm trying to think of a case where something like

```js
object.method(x => <div>{x}</div>)
//or
fn(object, x => <div>{x}</div>)
```

_isn't_ an iterator callback. So, at least optionally, feel like reporting on callbacks without keys would report the right situations without a lot of false positives.

**Comment by benmosher** [comment]
so my proposal would be `jsx-key` with `no-callback` option to disable the callback behavior if a given codebase has situations where it is over-reporting.

**Comment by yannickcr** [comment]
Fixed by #293
